File: YLPageScrollContentView.m
//
//  YLPageScrollContentView.m
//  YLPageScrollView
//
//  Created by Youngly on 2017/3/13.
//  Copyright © 2017年 YL. All rights reserved.
//

#import "YLPageScrollContentView.h"
#import "YLPageScrollTitleView.h"

@interface YLPageScrollContentView () <YLPageScrollTitleViewDelegate, UICollectionViewDataSource, UICollectionViewDelegate>

@property (nonatomic, strong) UICollectionView *collectionView;
@property (nonatomic, strong) UICollectionViewFlowLayout *layout;
@property (nonatomic, strong) NSArray<UIViewController *> *childVCs;
@property (nonatomic, strong) UIViewController *parentVC;
@property (nonatomic, strong) YLPageScrollViewAppreance *appreance;
@property (nonatomic, assign) CGFloat startOffsetX;
@property (nonatomic, assign) BOOL isForbidDelegate;
//@property (nonatomic, assign) BOOL isEndDecelerate;
@property (nonatomic, assign) NSUInteger indexWhenScrollToLeft;

@end


@implementation YLPageScrollContentView
static NSString * const cellId = @"UICollectionViewCell";

+ (instancetype)pageScrollContentViewWithFrame:(CGRect)frame childVCs:(NSArray *)childVCs parentVC:(UIViewController *)parentVC appreance:(YLPageScrollViewAppreance *)appreance
{
    return [[self alloc] initWithFrame:frame childVCs:childVCs parentVC:parentVC appreance:appreance];
}

- (instancetype)initWithFrame:(CGRect)frame childVCs:(NSArray *)childVCs parentVC:(UIViewController *)parentVC appreance:(YLPageScrollViewAppreance *)appreance
{
    if (self = [super initWithFrame:frame]) {
        
        self.backgroundColor = YLPageScrollViewRandomColor;
        self.isForbidDelegate = NO;
        self.indexWhenScrollToLeft = 0;
        
        self.childVCs = childVCs;
        self.appreance = appreance;
        self.parentVC = parentVC;
        
        [self addSubview:self.collectionView];
    }
    return self;
}

#pragma mark - UICollectionViewDataSource
- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section {
    return self.childVCs.count;
}

- (__kindof UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath
{
    UICollectionViewCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:cellId forIndexPath:indexPath];
    for (UIView *view in cell.subviews) {
        [view removeFromSuperview];
    }
    [cell addSubview:self.childVCs[indexPath.row].view];
    
    return cell;
}

#pragma mark - UICollectionViewDelegate
- (void)scrollViewDidScroll:(UIScrollView *)scrollView
{
    CGFloat contentOffsetX = scrollView.contentOffset.x;
    if (contentOffsetX == self.startOffsetX || self.isForbidDelegate) {
        return;
    }
    NSUInteger sourceIndex = 0;
    NSUInteger targetIndex = 0;
    CGFloat progress = 0;
    CGFloat scrollViewWidth = scrollView.bounds.size.width;
    if (contentOffsetX > self.startOffsetX) { // 向左滑动
        sourceIndex = (contentOffsetX) / scrollViewWidth;
        targetIndex = sourceIndex + 1;
        if (targetIndex > self.childVCs.count - 1) {
            targetIndex = self.childVCs.count - 1;
        }
        if (sourceIndex - self.indexWhenScrollToLeft == 1) {
            targetIndex = sourceIndex;
        }
//        NSLog(@"%f, %f", contentOffsetX, self.startOffsetX);
        NSLog(@"%lu, %lu", sourceIndex, targetIndex);
    } else { // 向右滑动
        
        targetIndex = contentOffsetX / scrollViewWidth;
        sourceIndex = targetIndex + 1;
        if (self.startOffsetX - contentOffsetX == scrollViewWidth) {
            sourceIndex = targetIndex;
        }
        NSLog(@"%lu, %lu", sourceIndex, targetIndex);
        NSLog(@"%f, %f", contentOffsetX, self.startOffsetX);
    }
    progress = ABS(contentOffsetX-self.startOffsetX) / scrollViewWidth;
    if (!self.isForbidDelegate && [self.delegate_ respondsToSelector:@selector(contentView:fromIndex:toIndex:progress:)]) {
        [self.delegate_ contentView:self fromIndex:sourceIndex toIndex:targetIndex progress:progress];
    }
    self.indexWhenScrollToLeft = sourceIndex;
}
- (void)scrollViewWillBeginDragging:(UIScrollView *)scrollView {
    self.isForbidDelegate = NO;
    self.startOffsetX = scrollView.contentOffset.x;
}
- (void)scrollViewDidEndDecelerating:(UIScrollView *)scrollView
{
    [self scrollViewDidEndScroll];
}
- (void)scrollViewWillEndDragging:(UIScrollView *)scrollView withVelocity:(CGPoint)velocity targetContentOffset:(inout CGPoint *)targetContentOffset
{
    NSLog(@"%s", __func__);
}
- (void)scrollViewDidEndDragging:(UIScrollView *)scrollView willDecelerate:(BOOL)decelerate
{
    if (!decelerate) {
        [self scrollViewDidEndScroll];
    } else {
        
    }
}
- (void)scrollViewDidEndScroll
{
    if ([self.delegate_ respondsToSelector:@selector(contentViewDidEndScroll:index:)]) {
        NSUInteger index = self.collectionView.contentOffset.x / self.frame.size.width;
        [self.delegate_ contentViewDidEndScroll:self index:index];
    }
}

#pragma mark - YLPageScrollTitleViewDelegate
- (void)pageScrollTitleViewDidSelected:(YLPageScrollTitleView *)titleView selectedIndex:(NSInteger)selectedIndex
{
    self.isForbidDelegate = YES;
    NSIndexPath *indexPath = [NSIndexPath indexPathForRow:selectedIndex inSection:0];
    [self.collectionView scrollToItemAtIndexPath:indexPath atScrollPosition:kNilOptions animated:NO];
}

#pragma mark - 懒加载
- (UICollectionViewFlowLayout *)layout {
    if (!_layout) {
        _layout = [UICollectionViewFlowLayout new];
        _layout.minimumLineSpacing = 0;
        _layout.minimumInteritemSpacing = 0;
        _layout.itemSize = self.bounds.size;
        _layout.scrollDirection = UICollectionViewScrollDirectionHorizontal;
    }
    return _layout;
}
- (UICollectionView *)collectionView {
    if (!_collectionView) {
        _collectionView = [[UICollectionView alloc] initWithFrame:self.bounds collectionViewLayout:self.layout];
        [_collectionView registerClass:[UICollectionViewCell class] forCellWithReuseIdentifier:cellId];
        _collectionView.dataSource = self;
        _collectionView.delegate = self;
        _collectionView.showsHorizontalScrollIndicator = NO;
        _collectionView.pagingEnabled = YES;
        _collectionView.bounces = NO;
    }
    return _collectionView;
}
@end

File: YLPageScrollView/YLPageScrollView/YLPageScrollContentView.swift
//
//  YLPageScrollContentView.swift
//  YLPageScrollView
//
//  Created by Youngly on 2017/3/12.
//  Copyright © 2017年 YL. All rights reserved.
//

import UIKit

private let cellId = "cellId"

protocol YLPageScrollContentViewDelegate : class {
    func contentView(_ contentView : YLPageScrollContentView, didEndScroll inIndex : Int)
    func contentView(_ contentView : YLPageScrollContentView, sourceIndex : Int, targetIndex : Int, progress : CGFloat)
}

class YLPageScrollContentView: UIView {

    weak var delegate : YLPageScrollContentViewDelegate?
    
    fileprivate var startOffsetX : CGFloat = 0
    fileprivate var childVCs : [UIViewController]
    fileprivate var parentVc : UIViewController
    fileprivate var isForbidDelegate : Bool = false
    fileprivate lazy var collectionView : UICollectionView = {
        let layout = UICollectionViewFlowLayout()
        layout.itemSize = self.bounds.size
        layout.minimumLineSpacing = 0
        layout.minimumInteritemSpacing = 0
        layout.scrollDirection = UICollectionViewScrollDirection.horizontal
        
        let collectionView = UICollectionView(frame: self.bounds, collectionViewLayout: layout)
        collectionView.dataSource = self
        collectionView.delegate = self
        collectionView.scrollsToTop = false
        collectionView.isPagingEnabled = true
        collectionView.bounces = false
        collectionView.showsVerticalScrollIndicator = false
        collectionView.showsHorizontalScrollIndicator = false
        collectionView.register(UICollectionViewCell.self, forCellWithReuseIdentifier: cellId)
        
        return collectionView
    }()
    
    init(frame: CGRect, childVCs: [UIViewController], parentVc:UIViewController) {
        
        self.childVCs = childVCs
        self.parentVc = parentVc
        
        super.init(frame: frame)
        backgroundColor = UIColor.randomColor()
        
        setupSubviews()
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
}

extension YLPageScrollContentView : UICollectionViewDataSource {
    
    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return childVCs.count
    }
    
    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: cellId, for: indexPath)
        for view in cell.contentView.subviews {
            view.removeFromSuperview()
        }
        cell.contentView.addSubview(childVCs[indexPath.row].view)
        
        return cell
    }
}

extension YLPageScrollContentView : UICollectionViewDelegate {
    
    func scrollViewDidEndDecelerating(_ scrollView: UIScrollView) {
        scrollViewDidEndScroll()
    }
    
    func scrollViewDidEndDragging(_ scrollView: UIScrollView, willDecelerate decelerate: Bool) {
        if !decelerate {
            scrollViewDidEndScroll()
        }
    }
    
    private func scrollViewDidEndScroll() {
        let index = Int(collectionView.contentOffset.x / collectionView.bounds.width)
        delegate?.contentView(self, didEndScroll: index)
    }
    
    func scrollViewWillBeginDragging(_ scrollView: UIScrollView) {
        isForbidDelegate = false
        startOffsetX = scrollView.contentOffset.x
    }
    
    func scrollViewDidScroll(_ scrollView: UIScrollView) {
        let contentOffsetX = scrollView.contentOffset.x
        guard contentOffsetX != startOffsetX && !isForbidDelegate else {
            return
        }
        
        var sourceIndex = 0
        var targetIndex = 0
        var progress : CGFloat = 0
        
        let collectionWidth = collectionView.bounds.width
        if contentOffsetX > startOffsetX { // 左滑动
            sourceIndex = Int(contentOffsetX / collectionWidth)
            targetIndex = sourceIndex + 1
            if targetIndex >= childVCs.count {
                targetIndex = childVCs.count - 1
            }
            progress = (contentOffsetX - startOffsetX) / collectionWidth
            if (contentOffsetX - startOffsetX) == collectionWidth {
                targetIndex = sourceIndex
            }
            
        } else { // 右滑动
            targetIndex = Int(contentOffsetX / collectionWidth)
            sourceIndex = targetIndex + 1
            progress = (startOffsetX - contentOffsetX) / collectionWidth
        }
        
        delegate?.contentView(self, sourceIndex: sourceIndex, targetIndex: targetIndex, progress: progress)
    }
}

// MARK: - YLPageScrollContentView
extension YLPageScrollContentView {
    fileprivate func setupSubviews() {
        for childVC in childVCs {
            parentVc.addChildViewController(childVC)
        }
        addSubview(collectionView)
    }
}

// MARK:- 遵守YLPageScrollTitleViewDelegate
extension YLPageScrollContentView : YLPageScrollTitleViewDelegate {
    func titleView(titleView: YLPageScrollTitleView, lastIndex: Int, selectedIndex: Int) {
        isForbidDelegate = true
        let indexPath = IndexPath(item: selectedIndex, section: 0)
        collectionView.scrollToItem(at: indexPath, at: .centeredHorizontally, animated: false)
    }
}
